### PR TITLE
Put subsuites with a few tests in one group

### PR DESCRIPTION
**Pull request by WeizhongX**
As we will need to restart the browser when starting a new subsuite, putting tests in one group can help reduce the number of restarts needed.

Keep an up limit of the size of such a group, so that we do not create a huge group that could potentially cause a worker to out live all other workers, thus increase the run time.

Bug: https://buganizer.corp.google.com/issues/341581718

### CODE REVIEW

**Review by jgraham — CHANGES_REQUESTED** [review]
The obvious feedback here is that the subsuite configuration should be allowed to decide, rather than having a hardcoded number. Even if there isn't per-subsuite configuration there should be a way to change this value from the command line.

**Comment by WeizhongX** [comment]
Having a command line argument sounds good to me.

**Comment by WeizhongX** [comment]
@jgraham, can you take another look at this? Thanks!

File: tools/wptrunner/wptrunner/wptcommandline.py [hunk 1]
```diff
@@ -150,6 +150,9 @@ def create_parser(product_choices=None):
     # TODO use an empty string argument for the default subsuite
     test_selection_group.add_argument("--subsuite", action="append", dest="subsuites",
                                       help="Subsuite names to run. Runs all subsuites when omitted.")
+    test_selection_group.add_argument("--small-subsuite-size", default=50, type=int,
+                                      help="Maximum number of tests a subsuite can have to be treated as small subsuite."
```
Comment (jonathan-j-lee) [inline tools/wptrunner/wptrunner/wptcommandline.py:154]:
```suggestion
                                      help="Maximum number of tests a subsuite can have to be treated as small subsuite. "
```

**Review by jonathan-j-lee — APPROVED** [review]

File: tools/wptrunner/wptrunner/testloader.py [hunk 2]
```diff
@@ -619,10 +619,30 @@ def new_group(self,
...
+            # needed. Tests have a different test type will still be put into
+            # different group though.
+            if self.in_one_group(subsuite, tests):
+                state["prev_group_key"] = (subsuite, test_type, [""])
+                group_metadata = self.group_metadata(state)
+                groups.append(TestGroup(deque(), subsuite, test_type, group_metadata))
+                for test in tests:
+                    group, _, _, metadata = groups[-1]
+                    group.append(test)
+                    test.update_metadata(metadata)
```
Comment (jgraham) [inline tools/wptrunner/wptrunner/testloader.py:643]: This bit is exactly like the loop below. Can you make the entire bit something like:

```py
in_one_group = self.in_one_group(subsuite, tests)
if in_one_group:
    state["prev_group_key"] = (subsuite, test_type, [""])
    group_metadata = self.group_metadata(state)
    groups.append(TestGroup(deque(), subsuite, test_type, group_metadata))

for test in tests:
    if not in_one_group and self.new_group(state, subsuite, test_type, group_metadata):
        group_metadata = self.group_metadata(state)
        groups.append(TestGroup(deque(), subsuite, test_type, group_metadata))

    group, _, _, metadata = groups[-1]
    group.append(test)
    test.update_metadata(metadata)
````

File: tools/wptrunner/wptrunner/testloader.py [hunk 3]
```diff
@@ -636,6 +656,11 @@ def tests_by_group(self, tests_by_type: TestsByType) -> Mapping[str, List[str]]:
         groups = defaultdict(list)
         state: MutableMapping[str, Any] = {}
         for (subsuite, test_type), tests in tests_by_type.items():
+            if self.in_one_group(subsuite, tests):
+                group_name = f"{subsuite}:/"
+                for test in tests:
+                    groups[group_name].append(test.id)
+                continue
```
Comment (jgraham) [inline tools/wptrunner/wptrunner/testloader.py:663]: Similar comment, or at least use an `else` rather than a `continue`.

**Review by jgraham — APPROVED** [review]
OK, but see some suggested small refactors to avoid repeating code that does the same thing.

**Review by jgraham — APPROVED** [review]

**Pull request merged**